### PR TITLE
Unit testing koans

### PR DESCRIPTION
**Pull request by iamvery**
Consider a koan like this:

```
defmodule AboutLists do
  ...
  think "Getting list length is a kernel feature" do
    a_list = [1, 2, 3]
    assert length(a_list) == __?
  end
  ...
end
```

It would be awesome if there was a way to provide the answers and then automatically test them. A potential syntax could be something like this:

```
defmodule AboutListsAnswers do
  ...
  enlighten "Getting list length is a kernel feature" do
    answer_for(__?, 3)
  end
  ...
end
```

The implementation details are kind of murky and would rely heavily on metaprogramming.
Some challenges off the top of my head:
- Need to support multiple `__?` in one `think` block
- Need to support `assert_?` as well
- Fail test suite if a `think` doesn't have an `answer`
- Have more than one test case for a `think` block (one for sad path, one for happy path)

### CODE REVIEW

**Comment by iamvery** [comment]
Attached the PR to this issue.. Getting close!

File: lib/koans/runner.ex [hunk 1]
```diff
@@ -32,7 +32,7 @@ defmodule Koans.Runner do
     failure(case, meditation)
     Koans.Formatter.failure_message(error, meditation, case)
     |> IO.puts
-    exit(:shutdown)
+    exit({:shutdown, 1})
```
Comment (iamvery) [inline lib/koans/runner.ex:35]: Added this so that CI build will fail if any examples are wrong.

File: about_asserts.exs [hunk 2]
```diff
@@ -14,6 +14,7 @@ defmodule AboutAsserts do
     ExUnit.Assertions.refute __?
   end
 
+  @tag :skip
```
Comment (iamvery) [inline about_asserts.exs]: This particular example doesn't require adding something, it's changing something... will need to figure out how to handle this. Possibly just rewrite the example.

File: about_asserts.exs [hunk 3]
```diff
@@ -36,6 +37,7 @@ defmodule AboutAsserts do
     assert_raise ExUnit.AssertionError, message, is_1_greater_than_2?
   end
 
+  @tag :skip
```
Comment (iamvery) [inline about_asserts.exs]: The solution doesn't handle `assert_?` yet. Will have to figure that out and remove all these skips.

File: answers/about_atoms.exs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  answer "Strings can be converted to atoms, and vice versa" do
+    [:atomized, "stringified"]
+  end
+
+  answer "Atoms are often used as keys, because they're faster than strings" do
+    ["Jay", "Jay"]
+  end
+
+  answer "Only atom keys may be accessed with dot syntax" do
+    ["Jay", map.name, "Jay"]
```
Comment (iamvery) [inline answers/about_atoms.exs:17]: I was _stunned_ that `map.name` worked here :mindblown:

**Comment by sebastiangeiger** [comment]
This looks awesome. I can't judge the metaprogramming but it obviously works and doesn't look very complicated.

I would say 🚢 .

**Pull request merged**